include/swift/AST/SourceFile.h

```diff
@@ -128,6 +130,14 @@ class SourceFile final : public FileUnit {
   /// The list of top-level declarations in the source file.
   std::vector<Decl *> Decls;
 
+  using ParserStatePtr =
+      std::unique_ptr<PersistentParserState, void (*)(PersistentParserState *)>;
+
+  /// Stores state that code completion needs to be able to resume parsing at
+  /// the code completion token in the file.
+  ParserStatePtr CodeCompletionParserState =
+      ParserStatePtr(/*ptr*/ nullptr, /*deleter*/ nullptr);
```

Comment:
`PersistentParserState` is (still) *not* only for code completion. Could you just use `ParserState` or so?

Reply:
@rintaro Hmm, I wanted to reflect the fact that the state is only set for source files registered for code completion with `setCodeCompletionPoint`, and that it corresponds to parser state recorded upon encountering a code completion token.

If we are to rename this, I would prefer we do it in conjunction with things like `setCodeCompletionPoint`, `isCodeCompletionFirstPass`, `CodeCompletionCallbacks` and possibly also `tok::code_complete`. What do you think?

Reply:
I see that `setCodeCompletionParserState()` is called only in code-completion, otherwise the parser uses its own `PersistentParserState`. But once it's set, the parser uses it to track `ScopeInfo` and `TopLevelContext`. So I still don't think `CodeCompletionParserState` is the right name for this.

> If we are to rename this, I would prefer we do it in conjunction with things like `setCodeCompletionPoint`, `isCodeCompletionFirstPass`, `CodeCompletionCallbacks` and possibly also `tok::code_complete`.

It's not clear to me what you are suggesting. Are you suggesting that if we rename `CodeCompletionParserState` to `ParserState` here, we should also rename, for example, `isCodeCompletionFirstPass` to `isParserFirstPass` or something?

Reply:
> It's not clear to me what you are suggesting. Are you suggesting that if we rename `CodeCompletionParserState` to `ParserState` here, we should also rename, for example, `isCodeCompletionFirstPass` to `isParserFirstPass` or something?

@rintaro Ah sorry, I thought your initial concern was about the fact that this is also used for other SourceKit operations (conforming method list + expr context info), so I was thinking that we could come up with a name that encompasses that family of operations.

> I see that `setCodeCompletionParserState()` is called only in code-completion, otherwise the parser uses its own `PersistentParserState`. But once it's set, the parser uses it to track `ScopeInfo` and `TopLevelContext`. So I still don't think `CodeCompletionParserState` is the right name for this.

Hmm, but the `TopLevelContext` being tracked is specific to code completion because on encountering the code completion token, we skip any remaining top-level decls such that the recorded `TopLevelContext` only has discriminators for the decls prior to the code completion token. 

For `ScopeInfo`, really we shouldn't need to track it all, as it should always be empty at the end of a parse. However we currently hang a couple of parser scopes off it when doing fast code completion inside a function body.

Overall I'm still reluctant to drop `CodeCompletion` from the name, as this is state needed by code completion's second pass to resume parsing before the code completion token, and shouldn't really be used for anything else.

I guess we should see what @slavapestov thinks :)

Reply:
@rintaro Do you still feel the same way about the naming? I'm asking because we now assert that `CodeCompletionParserState` has been set in the accessor, so it'll blow up when called on anything other than a parsed code completion file.

I talked with Slava offline and he suggested `DelayedParserState`, which seems like a good compromise if you still feel that we shouldn't put code completion in the name. What do you think?

Reply:
`DelayedParserState` sounds good!